source/slang/slang-check-decl.cpp

```diff
@@ -9308,8 +9308,160 @@ namespace Slang
+            // put the `ExpressionStmt` inside a `SeqStmt`.
+            // Since we found an init expression, we need to ensure
+            // we are adding to a constructor which can handle multiple
+            // statments since we will be adding `init` expressions to
+            // the constructor.
+            if (auto block = as<BlockStmt>(ctor->body); !as<SeqStmt>(block->body))
+            {
+                auto tmpExpr = block->body;
+                auto seqStmt = m_astBuilder->create<SeqStmt>();
+                seqStmt->stmts.add(tmpExpr);
+                block->body = seqStmt;
+            }
+            if (ctor->getParameters().getCount() != 0)
+                continue;
+            ctorDefault = ctor;
+        }
+        if (!ctorDefault)
+        {
+            ctorDefault = m_astBuilder->create<ConstructorDecl>();
+            auto ctorName = visitor->getName("$init");
+            ctorDefault->ownedScope = m_astBuilder->create<Scope>();
+            ctorDefault->ownedScope->containerDecl = ctorDefault;
+            ctorDefault->ownedScope->parent = visitor->getScope(structDecl);
+            ctorDefault->parentDecl = structDecl;
+            ctorDefault->loc = structDecl->loc;
+            ctorDefault->closingSourceLoc = ctorDefault->loc;
+            ctorDefault->nameAndLoc.name = ctorName;
+            ctorDefault->nameAndLoc.loc = ctorDefault->loc;
+            ctorDefault->returnType.type = visitor->calcThisType(makeDeclRef(structDecl));
+            auto body = m_astBuilder->create<BlockStmt>();
+            body->scopeDecl = m_astBuilder->create<ScopeDecl>();
+            body->scopeDecl->ownedScope = m_astBuilder->create<Scope>();
+            body->scopeDecl->ownedScope->parent = visitor->getScope(ctorDefault);
+            body->scopeDecl->parentDecl = ctorDefault;
+            body->scopeDecl->loc = ctorDefault->loc;
+            body->scopeDecl->closingSourceLoc = ctorDefault->loc;
+            body->closingSourceLoc = ctorDefault->closingSourceLoc;
+            ctorDefault->body = body;
+            body->body = m_astBuilder->create<SeqStmt>();
+            structDecl->addMember(ctorDefault);
+            ctorList.add(ctorDefault);
+        }
+
+        // It is possible a struct is a child of another struct/related-type.
+        // Since we are creating a new constructor function, it is possible
+        // a parent and a child to both have `__init()` constructors.
+        // As a result, this can cause a compile error when relying on the default
+        // constructor of an object because a child can use a parent object's
+        // constructor.
+        // To solve this issue and produce C++ like behavior we need to add overload
+        // ranks if not present already to prioritize the default constructor of a child.
+        // If an overload rank is manually set, we cannot assume any preference of
+        // constructor by the user and should ignore any special logic.
```

Comment:
I am not sure what this means.
Judging by the implementation, it sounds like,
```
If an overload rank is manually set, we need to respect the overload-rank from the user.
```
There is no point of having a comment if I have to deduce the meaning from the implementation.
The comment is to help the reader to understand the implementation; not the other way around.